#!/usr/bin/env bash
# 27-1-2023

# Megane step 3 - for the original code, 
# see: https://github.com/shohei-kojima/MEGAnE/wiki/Run-for-human-WGS 
## note that this step is optional

sif=megane_v1.0.1.beta.sif

# This step merges two vcf files generated by Megane in step 2
# Multi-allelic variants are removed by Megane in this step 
# This can be used for haplotype phasing 
sbatch -c 32 -t 04:00:00 -p cclake-himem --wrap="module load singularity; \
singularity exec megane_v1.0.1.beta.sif reshape_vcf \
-i 23-2-1_MEGANE_all_cichlids/jointcall_out/cichlid_all_MEI_jointcall.vcf.gz \
-a 23-2-1_MEGANE_all_cichlids/jointcall_out/cichlid_all_MEA_jointcall.vcf.gz \
-cohort_name cichlid_all"

# after I obtained the vcf file, I run these to get file w/o metadata for analysis in R
gunzip vcf_for_phasing/cichlid_all_MEI_biallelic.vcf.gz
gunzip vcf_for_phasing/cichlid_all_MEA_biallelic.vcf.gz
sed '/^##/ d' < vcf_for_phasing/cichlid_all_MEI_biallelic.vcf > vcf_for_phasing/cichlid_all_MEI_biallelic_no_meta.vcf
sed '/^##/ d' < vcf_for_phasing/cichlid_all_MEA_biallelic.vcf > vcf_for_phasing/cichlid_all_MEA_biallelic_no_meta.vcf

## change the encoding for PCA
sed -e 's#0/0#0#g' -e 's#0/1#1#g' -e 's#1/1#2#g' -e 's#0/.#0.5#g' -e 's#./0#0.5#g' -e 's#1/.#1.5#g' -e 's#./1#1.5#g' \
< vcf_for_phasing/cichlid_all_MEI_biallelic_no_meta.vcf > vcf_for_phasing/cichlid_all_MEI_biallelic_no_meta_encoded.vcf 
sed -e 's#0/0#0#g' -e 's#0/1#1#g' -e 's#1/1#2#g' -e 's#0/.#0.5#g' -e 's#./0#0.5#g' -e 's#1/.#1.5#g' -e 's#./1#1.5#g' \
< vcf_for_phasing/cichlid_all_MEA_biallelic_no_meta.vcf > vcf_for_phasing/cichlid_all_MEA_biallelic_no_meta_encoded.vcf

## to run: bash 23-1-27_megane_step3.sh
